### PR TITLE
Introduce new `GraqlConverter` interface

### PR DESCRIPTION
**Pull request by aelred**
# Why is this PR needed?

gRPC will need a way to convert Graql results into streams of gRPC data. Currently, we only support converting results into streams of *strings*.

# What does the PR do?

We introduce a new `GraqlConverter` interface that `Printer` extends.

# Does it break backwards compatibility?

No. I consider only `Printer#graqlString(Object)` and `Query#resultsString(Printer)` to be part of our public `Printer` API and I have deprecated both methods.

There are other "public" methods in this PR that I've renamed, but I don't consider them part of our API.

# List of future improvements not on this PR

Actually using this new abstraction! gRPC will likely use it, but it may come in handy in other places (for example, the `JacksonPrinter` may be able to return something more descriptive than `String`).

### CODE REVIEW

File: grakn-engine/src/main/java/ai/grakn/engine/controller/response/ExplanationBuilder.java [hunk 1]
```diff
@@ -59,7 +59,8 @@
                         .flatMap(unifier -> rewrittenQuery.materialise(answer.unify(unifier)))
                         .findFirst().orElse(new QueryAnswer());
             }
-            explanation.add((Answer) printer.graqlString(false, inferredAnswer));
+            // TODO: this looks like a bug: the return type is not `Answer`
```
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/controller/response/ExplanationBuilder.java]: Not a bug just poor design. In the context of the explanation builder the return type is actually always an answer.

File: grakn-core/src/main/java/ai/grakn/graql/GraqlConverter.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+/**
+ * An interface for converting objects in Graql responses into other formats.
+ *
+ * @param <Builder> An intermediate builder type that can be changed into a {@link T}
+ * @param <T> The type to convert into
+ *
+ * @author Felix Chapman
+ */
+public interface GraqlConverter<Builder, T> {
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/GraqlConverter.java]: The distinction between `GraqlConverter` and `Printer` is kind of vague. Maybe even just changing the to be more explicit to something like:

```
An interface for converting any objects (e.g. Answers, Concepts, etc . . . ) in Graql responses into specific Types `T`.
```

File: grakn-core/src/main/java/ai/grakn/graql/GraqlConverter.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+import javax.annotation.CheckReturnValue;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * An interface for converting objects in Graql responses into other formats.
+ *
+ * @param <Builder> An intermediate builder type that can be changed into a {@link T}
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/GraqlConverter.java]: Struggled to understand why this was needed. Not sure how to improve on this.

File: grakn-engine/src/main/java/ai/grakn/engine/controller/response/ExplanationBuilder.java (hunk 1 above)
Comment (aelred) [reply grakn-engine/src/main/java/ai/grakn/engine/controller/response/ExplanationBuilder.java]: Thanks! I get it. So `JacksonPrinter` returns `Object` and therefore sometimes it is an `Answer`. In this context, we know that it must be.

@fppt suggested we can make this a bit clearer by making this method take a `Printer<Object>` or `JacksonPrinter` explicitly.

**Pull request merged**